search/ranker.cpp

```diff
@@ -901,9 +888,6 @@ void Ranker::ProcessSuggestions(vector<RankerResult> & vec) const
         {
           ++added;
         }
-
-        i = vec.erase(i);
```

Comment:
Why is it not removed now?

Reply:
This fixes the third bullet in #1376. `vec` is a shallow copy of the search results, so erasing the element means there's an autocomplete match for the feature but not a result the user can select.

The attached image shows an example. In the PR version (top) "S Quebec Street" is present both as an autocomplete hint and a result. In the bottom (current), it's only a suggestion. That forces the user to click the suggestion and rerun the search.

![2021-12-08-082314_491x772_scrot](https://user-images.githubusercontent.com/4681424/145234684-d93720fd-1c04-46ce-829e-596f58bf9a5a.png)

Reply:
Wait, but then the result (if it is NOT a street, like in #1376)  should be removed from the suggestion: existing logic should be inverted. Why user needs the result (not a street) in the suggestion? The case for streets is to type the full address with fewer clicks. That's why they need a suggestion (and you're right, they shouldn't be displayed in results).

Reply:
I think it already works that way - suggestions require the `RankerResult` to be a country, city, or street:
```
matcher.cpp:879     if (type == ftypes::LocalityType::Country || type == ftypes::LocalityType::City || r.IsStreet())
```
But I'm not sure how that intersects with #1376. Maybe there's a street that's also called Carrefour that was converted to a suggestion, and the search never found the SUBPOI?

Reply:
Well, this change is not about #1376.
The logic of "i = vec.erase(i);" - If some street (country, city) result goes to the suggestion part, it will NOT go to the results list. Actually, I can't say for sure is it good or bad, so I also agree to change this logic and test.
Now we have "South Quebec Street" as a suggestion AND as a result (see highway-secondary).

Reply:
What about this approach?
1. Query matches the same result and suggestion partially: display both result and suggestion in all cases except one: do not display any street type as a result (people do not usually search for streets alone).
2. Query fully matches the same result and suggestion and doesn't have a space at the end: the same as (1) above.
3. Query fully matches the same result and suggestion and have a space at the end: display the result only (even if it's a street type).

Reply:
At least for the US, I think it's important to present streets as results.

As mentioned in #1263, lots of people have trouble when the address they entered doesn't exist. In the US, there are many areas where roads are mapped fully but street numbers aren't. Returning the street as a result is an easy fallback to get a user close to the right place.

That's what I was thinking in including the suggestions as results. (It'd obviously be nicer if all buildings existed and had addresses, but that's a much larger project.)

Reply:
Users will see their street as a suggestion (but not in results, saving potential space for another relevant result or suggestion). And they click on the suggested street. And it will appear in results _after_ the click was done (and disappear from suggestions).

Reply:
The main idea with suggestions is to save the number of clicks and avoid potential mistypes. The most popular pattern is to type several first street letters, then click a suggestion, and then type the house number.
In the US, however, people usually type the number first and then type the street name.

If there is a house with a number, then it should be in results (and no street alone, of course). If there is no house, then only street should be displayed in results (but not in suggestion in both cases).

Reply:
- Turned out that this is suggestion only logic is confusing a bit. Also suggestions have limits.
- If we search for street only with suggestions, we don't have fuzzy (partial) matching in this case.

I'd like to check how it will go with street results. And I like it so far.